agent/consul/state/catalog.go

```diff
@@ -2363,79 +2413,125 @@ func checkSessionsTxn(tx *memdb.Txn, hc *structs.HealthCheck) ([]*sessionCheck,
+	case structs.TerminatingGateway:
+		gatewayServices, err = s.terminatingConfigGatewayServices(tx, gatewayID, conf, entMeta)
 	}
-	if cfg, ok := c.(*structs.TerminatingGatewayConfigEntry); ok && cfg != nil {
-		if reflect.DeepEqual(cfg.Services, entry.Services) {
-			// Services are the same, nothing to update
-			return nil
-		}
+	// Return early if there is an error OR we don't have any services to update
+	if err != nil || len(gatewayServices) == 0 {
+		return err
 	}
 
 	// Delete all associated with gateway first, to avoid keeping mappings that were removed
-	if _, err := tx.DeleteAll(terminatingGatewayServicesTableName, "gateway", structs.NewServiceID(entry.Name, entMeta)); err != nil {
+	if _, err := tx.DeleteAll(gatewayServicesTableName, "gateway", structs.NewServiceID(conf.GetName(), entMeta)); err != nil {
 		return fmt.Errorf("failed to truncate gateway services table: %v", err)
 	}
 
-	gatewayID := structs.NewServiceID(entry.Name, &entry.EnterpriseMeta)
-	for _, svc := range entry.Services {
+	for _, svc := range gatewayServices {
 		// If the service is a wildcard we need to target all services within the namespace
-		if svc.Name == structs.WildcardSpecifier {
-			if err := s.updateTerminatingGatewayNamespace(tx, gatewayID, svc, entMeta); err != nil {
+		if svc.Service.ID == structs.WildcardSpecifier {
+			if err := s.updateGatewayNamespace(tx, idx, svc, entMeta); err != nil {
 				return fmt.Errorf("failed to associate gateway %q with wildcard: %v", gatewayID.String(), err)
 			}
 			// Skip service-specific update below if there was a wildcard update
 			continue
 		}
 
-		// Check if the non-wildcard service is already associated with a gateway
-		existing, err := s.serviceTerminatingGateway(tx, svc.Name, &svc.EnterpriseMeta)
-		if err != nil {
-			return fmt.Errorf("gateway service lookup failed: %s", err)
-		}
-		if gs, ok := existing.(*structs.GatewayService); ok && gs != nil {
-			// Only return an error if the stored gateway does not match the one from the config entry
-			if !gs.Gateway.Matches(&gatewayID) {
-				return fmt.Errorf("service %q is associated with different gateway, %q", gs.Service.String(), gs.Gateway.String())
-			}
-		}
-
 		// Since this service was specified on its own, and not with a wildcard,
 		// if there is an existing entry, we overwrite it. The service entry is the source of truth.
 		//
 		// By extension, if TLS creds are provided with a wildcard but are not provided in
 		// the service entry, the service does not inherit the creds from the wildcard.
+		err = s.updateGatewayService(tx, idx, svc)
+		if err != nil {
+			return err
+		}
+	}
+
+	if err := indexUpdateMaxTxn(tx, idx, gatewayServicesTableName); err != nil {
+		return fmt.Errorf("failed updating gateway-services index: %v", err)
+	}
+	return nil
+}
+
+func (s *Store) ingressConfigGatewayServices(tx *memdb.Txn, gateway structs.ServiceID, conf structs.ConfigEntry, entMeta *structs.EnterpriseMeta) (structs.GatewayServices, error) {
```

Comment:
If you return `nil` or `len()==0` to mean "nope, no update" how would someone refresh a definition to legitimately have no gateway services? Do you need to change the return contract to something like `(svcs structs.GatewayServices, shouldUpdate bool, err error)`?

Reply:
Hmm true. I wonder whether removing the `len()==0` check in `updateGatewayServices` is sufficient here though. 🤔 

`ingressConfigGatewayServices` should always construct the set of gateway-services that a config entry represents, and as such a 0 length array should be valid.

Reply:
Ahh I see, my comment above, plus adding an error the the switch statement in `updateGatewayServices` might be the ticket.

Reply:
Pushed a commit that deals with this, plus additional test case.